services/snapcraft/snapcraft-version.service.js

```diff
@@ -0,0 +1,91 @@
+import Joi from 'joi'
+import { BaseJsonService, NotFound, pathParams, queryParam } from '../index.js'
+import { renderVersionBadge } from '../version.js'
+
+const queryParamSchema = Joi.object({
+  arch: Joi.string(),
+})
+
+const versionSchema = Joi.object({
+  'channel-map': Joi.array()
+    .items(
+      Joi.object({
+        channel: Joi.object({
+          architecture: Joi.string().required(),
+          risk: Joi.string().required(),
+          track: Joi.string().required(),
+        }),
+        version: Joi.string().required(),
+      }).required(),
+    )
+    .min(1)
+    .required(),
+}).required()
+
+export default class SnapcraftVersion extends BaseJsonService {
+  static category = 'version'
+
+  static route = {
+    base: 'snapcraft/v',
+    pattern: ':package/:track/:risk',
+    queryParamSchema,
+  }
+
+  static defaultBadgeData = { label: 'snapcraft' }
+
+  static openApi = {
+    '/snapcraft/v/{package}/{track}/{risk}': {
+      get: {
+        summary: 'Snapcraft version',
+        parameters: [
+          ...pathParams(
+            { name: 'package', example: 'chromium' },
+            { name: 'track', example: 'latest' },
+            { name: 'risk', example: 'stable' },
+          ),
+          queryParam({
+            name: 'arch',
+            example: 'amd64',
+            description:
+              'Architecture, When not specified, this will default to `amd64`.',
+          }),
+        ],
+      },
+    },
+  }
+
+  async handle({ package: packageName, track, risk }, { arch = 'amd64' }) {
+    const parsedData = await this._requestJson({
+      schema: versionSchema,
+      options: {
+        headers: { 'Snap-Device-Series': 16 },
+      },
+      url: `https://api.snapcraft.io/v2/snaps/info/${packageName}`,
+      httpErrors: {
+        404: 'package not found',
+      },
+    })
+
+    const channelMap = parsedData['channel-map']
+    let filteredChannelMap = channelMap.filter(
+      ({ channel }) => channel.architecture === arch,
+    )
+    if (filteredChannelMap.length === 0) {
+      throw new NotFound({ prettyMessage: 'arch not found' })
+    }
+    filteredChannelMap = channelMap.filter(
+      ({ channel }) => channel.track === track,
+    )
+    if (filteredChannelMap.length === 0) {
+      throw new NotFound({ prettyMessage: 'track not found' })
+    }
+    filteredChannelMap = channelMap.filter(
+      ({ channel }) => channel.risk === risk,
+    )
+    if (filteredChannelMap.length === 0) {
+      throw new NotFound({ prettyMessage: 'risk not found' })
+    }
```

Comment:
This code looks wrong to me. These filters should be combined with AND logic, but we're filtering `channelMap`, assigning the result to `filteredChannelMap`and then essentially throwing away that result and overwriting `filteredChannelMap`.

I think you really want to be doing

```js
filteredChannelMap = filteredChannelMap.filter(...
```

for the second and third pass.

What would really help here would be if we split this filtering logic out into a `transform()` function and write a few unit tests (in a spec.js) to make sure this is working as expected.

You might look at something like

`services/docker/docker-version.spec.js`
or
`services/crates/crates-base.spec.js`

as a point of inspiration.

Reply:
My bad, i copy-pasted without noticing i forgot to change to filteredChannelMap

Fixed filter logic to AND with https://github.com/badges/shields/pull/9976/commits/b54f1e649a5ac61c67694c648034ed211e64eb2e
Moved the filter logic into transform with https://github.com/badges/shields/pull/9976/commits/75f8ac763b8c062464510f6ce90c44a28356a34c
And most impotent, added testing with https://github.com/badges/shields/pull/9976/commits/29dd792cd42c823811b4614c75b0d1ffdaac8745
This test file is really called for.